### PR TITLE
KAFKA-13572 Fix negative preferred replica imbalanced count metric

### PR DESCRIPTION
**Pull request by ocadaruma**
* Currently, `preferredReplicaImbalanceCount` calculation has a race that becomes negative when topic deletion is initiated simultaneously.
    * Please refer [KAFKA-13572's comment](https://issues.apache.org/jira/browse/KAFKA-13572?focusedCommentId=17566872&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-17566872) for the details
* This PR addresses the problem by fixing `cleanPreferredReplicaImbalanceMetric` to be called only once per topic-deletion procedure

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: core/src/main/scala/kafka/controller/ControllerContext.scala
Comment (showuon) [inline core/src/main/scala/kafka/controller/ControllerContext.scala]: 1. Could we add a comment here to mention why we need `diff` here?
2. I don't think the parameter name `topics` is clear enough. Could we update it? Ex: topics -> `newTopicsToBeDeleted` or `topicToBeAddedIntoDeletionList` or something else?

File: core/src/main/scala/kafka/controller/ControllerContext.scala
Comment (showuon) [inline core/src/main/scala/kafka/controller/ControllerContext.scala]: Since we've calculated the diff, we can directly add the `newlyDeletedTopics` to avoid unnecessary duplication, right?

File: core/src/main/scala/kafka/controller/ControllerContext.scala
Comment (showuon) [reply core/src/main/scala/kafka/controller/ControllerContext.scala]: One question: After this change, `topic-A` in your example in JIRA will be only entered into `cleanPreferredReplicaImbalanceMetric` once, right? So, what if when the first time entering `cleanPreferredReplicaImbalanceMetric`, it haven't run topic-A's deletion procedure to mark topic-A's all partitions as Offline (Leader = -1) here: https://github.com/apache/kafka/blob/3.2.0/core/src/main/scala/kafka/controller/ReplicaStateMachine.scala#L368 ? I mean, since we only decrement `preferredReplicaImbalanceCount` if topic-A has NoLeader and `!hasPreferredLeader(replicaAssignment, leadershipInfo)` evaluates to true, right? So, what if the topic-A haven't reached this condition and we entering `cleanPreferredReplicaImbalanceMetric`? Will that cause some partitions count should be decremented but they are not? Or I should ask, will that case happen?

**Review by showuon — COMMENTED** [review]
@ocadaruma , thanks for the PR. Left some comments.

File: core/src/main/scala/kafka/controller/ControllerContext.scala
Comment (ocadaruma) [reply core/src/main/scala/kafka/controller/ControllerContext.scala]: (I'm not 100% sure if I get your concern. If my answer doesn't make sense, please tell me)

In the first place, the purpose of `cleanPreferredReplicaImbalanceMetric` is to "unmark" partitions which has imbalanced leader.

So, below is the expected behavior (let's say topic-A has single partition):
- Initiate topic-A deletion
- If topic-A-0 has imbalanced leader (e.g. lost leadership once, replica-reassignment, became offline etc) here,  `preferredReplicaImbalanceCount` contains the count for it here, so we have to decrement it because the partition is to be deleted
  * `!hasPreferredLeader` evaluates to true so decremented as expected
- Otherwise, we don't have to decrement the value because `preferredReplicaImbalanceCount` doesn't contain the imbalance count for `topic-A-0`.

"A partition should be decremented from preferredReplicaImbalanceCount on topic-deletion only if the partition is actually imbalanced at the topic-deletion timing" is the invariant that topic-deletion procedure must satisfy.

So, essentially, the problem is that topic-A-0 could be marked as "imbalanced" wrongly when it becomes offline during deletion-procedure.

> So, what if the topic-A haven't reached this condition and we entering cleanPreferredReplicaImbalanceMetric? Will that cause some partitions count should be decremented but they are not?

"topic-A haven't reached this condition" means the partition wasn't imbalanced actually, so we do not have to decrement the count.

Comment (showuon) [reply core/src/main/scala/kafka/controller/ControllerContext.scala]: @ocadaruma , thanks for the explanation. So, do you mean:

Before the PR, topic-A partition 0 is in preferred Leader, so after 1st time entering `cleanPreferredReplicaImbalanceMetric`, it won't do anything, which is expected, because ithe `preferredReplicaImbalanceCount` is not incremented before. But after 1st `topic-A` deletion, the leader will be changed to NoLeader (or other replica), and now, if we have 2nd chance entering  `cleanPreferredReplicaImbalanceMetric`, it'll decrement it, which is unexpected. Is my understanding correct?

File: core/src/test/scala/unit/kafka/controller/ControllerContextTest.scala
Comment (showuon) [inline core/src/test/scala/unit/kafka/controller/ControllerContextTest.scala]: It's a little confusing we sometimes call `tp1's` deletion, sometimes call `tp3's topic` deletion. Could we create 2 variables for these 2 topics? ex: `topicA`, `topicB`? So, in this test case, we can make it much clear like this:
```
context.queueTopicDeletion(Set(topicA))

// Initiate topicB's deletion before topicA's deletion completes.
...
```

**Review by showuon — COMMENTED** [review]
@ocadaruma , left a minor comment. Overall LGTM! Do you think we should backport this fix to v3.2?

File: core/src/main/scala/kafka/controller/ControllerContext.scala
Comment (ocadaruma) [reply core/src/main/scala/kafka/controller/ControllerContext.scala]: @showuon Correct.

**Comment by ocadaruma** [comment]
> Do you think we should backport this fix to v3.2?

I think it's worth but honestly I'm not sure about the criteria for backport so I would like to leave it to committer's decision.

File: core/src/main/scala/kafka/controller/ControllerContext.scala
Comment (showuon) [reply core/src/main/scala/kafka/controller/ControllerContext.scala]: I understand now. Thanks.

**Review by showuon — APPROVED** [review]
LGTM!

**Comment by showuon** [comment]
Let's wait for the jenkins build completes. Thanks.

**Comment by showuon** [comment]
All tests passed.

**Comment by showuon** [comment]
backport to 3.3 and 3.2. cc @jsancio @mumrah

**Pull request merged**
